Thanks for providing these changes, Dario.

I have tested those in a python3 environment/agent, and here is the outcome for workflowCompletion.py (working fine!):
```
cmst1@vocms0192:/data/srv/wmagent/current $ python3 apps/wmagentpy3/bin/adhoc-scripts/workflowCompletion.py -w amaltaro_TaskChain_PUMCRecyc_Agent152_Val_210817_124215_6631 -c cmsweb-testbed.cern.ch -r alancc7-cloud1.cern.ch
apps/wmagentpy3/bin/adhoc-scripts/workflowCompletion.py:58: DeprecationWarning: key_file, cert_file and check_hostname are deprecated, use a custom context instead.
  return HTTPSConnection(host, key_file=self.key, cert_file=self.cert)
==> amaltaro_TaskChain_PUMCRecyc_Agent152_Val_210817_124215_6631	(status: completed)
InputDataset:
    /RelValQCD_Pt_600_800_14/CMSSW_11_2_0_pre8-112X_mcRun3_2024_realistic_v10_forTrk-v1/GEN-SIM (total input lumis: 180)
OutputDatasets:
    /RelValQCD_Pt_600_800_14/CMSSW_11_2_0_pre8-DigiPU_2024PU_TaskChain_PUMCRecyc_Agent152_Val_Alanv11-v11/GEN-SIM-DIGI-RAW (lumis: 180, lumi completion: 1.0)
    /RelValQCD_Pt_600_800_14/CMSSW_11_2_0_pre8-RecoPU_2024PU_TaskChain_PUMCRecyc_Agent152_Val_Alanv11-v11/DQMIO (lumis: 180, lumi completion: 1.0)
    /RelValQCD_Pt_600_800_14/CMSSW_11_2_0_pre8-RecoPU_2024PU_TaskChain_PUMCRecyc_Agent152_Val_Alanv11-v11/MINIAODSIM (lumis: 180, lumi completion: 1.0)
    /RelValQCD_Pt_600_800_14/CMSSW_11_2_0_pre8-RecoPU_2024PU_TaskChain_PUMCRecyc_Agent152_Val_Alanv11-v11/GEN-SIM-RECO (lumis: 180, lumi completion: 1.0)
```

while we have problems with the validate script:
```
amaltaro@vocms0192:/data/amaltaro/July2020_Val $ python3 --version
Python 3.8.2
amaltaro@vocms0192:/data/amaltaro/July2020_Val $ python3 validate-test-wfs.py -w amaltaro_TaskChain_PUMCRecyc_Agent152_Val_210817_124215_6631 -c cmsweb-testbed.cern.ch -r alancc7-cloud1.cern.ch
validate-test-wfs.py:68: DeprecationWarning: key_file, cert_file and check_hostname are deprecated, use a custom context instead.
  return HTTPSConnection(host, key_file=self.key, cert_file=self.cert)
Traceback (most recent call last):
  File "validate-test-wfs.py", line 713, in <module>
    sys.exit(main())
  File "validate-test-wfs.py", line 673, in main
    rucioToken = getRucioToken(rucioUrl)
  File "validate-test-wfs.py", line 132, in getRucioToken
    token = getContent(rucioAuth, headers=[("X-Rucio-Account", rucioAcct)])
  File "validate-test-wfs.py", line 100, in getContent
    output = response.headers.getheader('X-Rucio-Auth-Token')
AttributeError: 'HTTPMessage' object has no attribute 'getheader'
```

--- a/bin/adhoc-scripts/workflowCompletion.py
+++ b/bin/adhoc-scripts/workflowCompletion.py
@@ -6,26 +6,45 @@ completion, so the ration of input lumis vs output lumis.
 from __future__ import print_function, division
 
 import argparse
-import httplib
 import json
 import os
 import pwd
 import sys
 import urllib
-import urllib2
-from urllib2 import HTTPError, URLError
+
+try:
+    # python2
+    import urllib2
+    HTTPError, URLError = urllib2.HTTPError, urllib2.URLError
+    build_opener = urllib2.build_opener
+    HTTPSHandler = urllib2.HTTPSHandler
+    import urllib
+    urlencode = urllib.urlencode
+    import httplib
+    HTTPSConnection = httplib.HTTPSConnection
+except:
+    # python3
+    import urllib.error
+    HTTPError, URLError = urllib.error.HTTPError, urllib.error.URLError
+    import urllib.request
+    build_opener = urllib.request.build_opener
+    HTTPSHandler = urllib.request.HTTPSHandler
+    import urllib.parse
+    urlencode = urllib.parse.urlencode
+    import http.client
+    HTTPSConnection = http.client.HTTPSConnection
 
 # ID for the User-Agent
 CLIENT_ID = 'workflowCompletion::python/%s.%s' % sys.version_info[:2]
 
 
-class HTTPSClientAuthHandler(urllib2.HTTPSHandler):
+class HTTPSClientAuthHandler(HTTPSHandler):
     """
     Basic HTTPS class
     """
 
     def __init__(self, key, cert):
-        urllib2.HTTPSHandler.__init__(self)
+        HTTPSHandler.__init__(self)
         self.key = key
         self.cert = cert
 
@@ -36,7 +55,7 @@ class HTTPSClientAuthHandler(urllib2.HTTPSHandler):
         return self.do_open(self.getConnection, req)
 
     def getConnection(self, host, timeout=290):
-        return httplib.HTTPSConnection(host, key_file=self.key, cert_file=self.cert)
+        return HTTPSConnection(host, key_file=self.key, cert_file=self.cert)
 
 
 def getX509():
@@ -58,7 +77,7 @@ def getContent(url, params=None):
     certFile, keyFile = getX509()
     client = '%s (%s)' % (CLIENT_ID, os.environ.get('USER', ''))
     handler = HTTPSClientAuthHandler(keyFile, certFile)
-    opener = urllib2.build_opener(handler)
+    opener = build_opener(handler)
     opener.addheaders = [("User-Agent", client),
                          ("Accept", "application/json")]
 
@@ -113,7 +132,7 @@ def handleDBS(reqmgrOutDsets, cmswebUrl):
 
     dbsOutput = {}
     for dataset in reqmgrOutDsets:
-        fullUrl = dbsUrl + "filesummaries?" + urllib.urlencode({'dataset': dataset})
+        fullUrl = dbsUrl + "filesummaries?" + urlencode({'dataset': dataset})
         data = json.loads(getContent(fullUrl))
         if data:
             dbsOutput[dataset] = data[0]['num_lumi']

--- a/test/data/ReqMgr/validate-test-wfs.py
+++ b/test/data/ReqMgr/validate-test-wfs.py
@@ -1,21 +1,41 @@
 #!/usr/bin/env python
 from __future__ import print_function
+from future.utils import viewitems
 
 import argparse
 import getpass
-import httplib
 import json
 import os
 import pwd
 import sys
-import urllib
-import urllib2
 from collections import OrderedDict
 from textwrap import TextWrapper
-from urllib import quote_plus
-from urllib2 import HTTPError, URLError
 
-from future.utils import viewitems
+try:
+    # python2
+    import urllib2
+    HTTPError = urllib2.HTTPError
+    URLError = urllib2.URLError
+    build_opener = urllib2.build_opener
+    HTTPSHandler = urllib2.HTTPSHandler
+    import urllib
+    urlencode = urllib.urlencode
+    quote_plus = urllib.quote_plus
+    import httplib
+    HTTPSConnection = httplib.HTTPSConnection
+except:
+    # python3
+    import urllib.error
+    HTTPError = urllib.error.HTTPError, 
+    URLError = urllib.error.URLError
+    import urllib.request
+    build_opener = urllib.request.build_opener
+    HTTPSHandler = urllib.request.HTTPSHandler
+    import urllib.parse
+    urlencode = urllib.parse.urlencode
+    quote_plus = urllib.parse.quote_plus
+    import http.client
+    HTTPSConnection = http.client.HTTPSConnection
 
 # table parameters
 SEPARATELINE = "|" + "-" * 51 + "|"
@@ -28,13 +48,13 @@ CLIENT_ID = 'validate-test-wfs/1.2::python/%s.%s' % sys.version_info[:2]
 cachedDqmgui = None
 
 
-class HTTPSClientAuthHandler(urllib2.HTTPSHandler):
+class HTTPSClientAuthHandler(HTTPSHandler):
     """
     Basic HTTPS class
     """
 
     def __init__(self, key, cert):
-        urllib2.HTTPSHandler.__init__(self)
+        HTTPSHandler.__init__(self)
         self.key = key
         self.cert = cert
 
@@ -45,7 +65,7 @@ class HTTPSClientAuthHandler(urllib2.HTTPSHandler):
         return self.do_open(self.getConnection, req)
 
     def getConnection(self, host, timeout=290):
-        return httplib.HTTPSConnection(host, key_file=self.key, cert_file=self.cert)
+        return HTTPSConnection(host, key_file=self.key, cert_file=self.cert)
 
 
 def getX509():
@@ -67,7 +87,7 @@ def getContent(url, params=None, headers=None):
     certFile, keyFile = getX509()
     client = '%s (%s)' % (CLIENT_ID, os.environ.get('USER', ''))
     handler = HTTPSClientAuthHandler(keyFile, certFile)
-    opener = urllib2.build_opener(handler)
+    opener = build_opener(handler)
     if headers:
         opener.addheaders = headers
     else:
@@ -77,7 +97,8 @@ def getContent(url, params=None, headers=None):
     try:
         response = opener.open(url, params)
         if "auth/x509" in url:
-            output = response.headers.getheader('X-Rucio-Auth-Token')
+            output = response.headers.get('X-Rucio-Auth-Token', '')  # option 1
+            # output = response.getheader('X-Rucio-Auth-Token')  # option 2
         else:
             output = response.read()
     except HTTPError as e:
@@ -190,7 +211,7 @@ def reader(stream):
     """
     Home-made function to consume newline delimited json streaming data
     """
-    for line in stream.split("\n"):
+    for line in stream.split(b"\n"):
         if line:
             yield json.loads(line)
 
@@ -230,12 +251,12 @@ def getDbsInfo(dataset, baseUrl):
     dbsOutput = {}
     dbsOutput[dataset] = {"blockorigin": [], "filesummaries": [], "outputconfigs": []}
     for api in dbsOutput[dataset]:
-        fullUrl = dbsUrl + api + "?" + urllib.urlencode({'dataset': dataset})
+        fullUrl = dbsUrl + api + "?" + urlencode({'dataset': dataset})
         data = json.loads(getContent(fullUrl))
         dbsOutput[dataset][api] = data
 
     # Separate query for prep_id, since we want any access_type
-    fullUrl = dbsUrl + "datasets?" + urllib.urlencode({'dataset': dataset})
+    fullUrl = dbsUrl + "datasets?" + urlencode({'dataset': dataset})
     fullUrl += "&dataset_access_type=*&detail=True"
     data = json.loads(getContent(fullUrl))
     dbsOutput[dataset]["prep_id"] = data[0]["prep_id"] if data else ''
@@ -338,7 +359,7 @@ def compareSpecial(d1, d2, key=None):
 
         if key == 'PNN':
             for dset in d2:
-                for block, value in d2[dset].iteritems():
+                for block, value in viewitems(d2[dset]):
                     if isinstance(value, dict):
                         if d1[dset][block][key] != d2[dset][block][key]:
                             return outcome
@@ -382,7 +403,7 @@ def twClosure(replace_whitespace=False,
             if reCall:
                 output += '\n'
             ind += '    '
-            for key, value in obj.iteritems():
+            for key, value in viewitems(obj):
                 output += "%s%s: %s" % (ind,
                                         ''.join(twr.wrap(key)),
                                         twEnclosed(value, ind, reCall=True))
@@ -592,7 +613,7 @@ def validateAll(reqmgrInputDset, couchInfo, rucioInfo, dbsInfo):
     print('|' + ' ' * 25 + '| CouchDB | Rucio  | DBS  |')
     print(SEPARATELINE)
 
-    compRes = compareLists(couchInfo.keys(), rucioInfo.keys(), dbsInfo.keys())
+    compRes = compareLists(list(couchInfo.keys()), list(rucioInfo.keys()), list(dbsInfo.keys()))
     print('| Same dataset name       | {compRes:7s} | {compRes:6s} | {compRes:4s} |'.format(compRes=compRes))
 
     compRes = compareLists(couchInfo, rucioInfo, dbsInfo, key='dsetSize')

